swarm/swap_test.go

```diff
@@ -0,0 +1,468 @@
+			}
+		}
+	}
+
+	//now iterate the whole map
+	//and check that every node k has the same
+	//balance with a peer as that peer with the node,
+	//but in inverted signs
+
+	//iterate the map
+	success := true
+	for k, mapForK := range balancesMap {
+		//iterate the submap
+		for n, balanceKwithN := range mapForK {
+			//iterate the main map again
+			mapForSubK := balancesMap[n]
+			log.Trace(fmt.Sprintf("balance of %s with %s: %d", k.TerminalString(), n.TerminalString(), balanceKwithN))
+			log.Trace(fmt.Sprintf("balance of %s with %s: %d", n.TerminalString(), k.TerminalString(), mapForSubK[k]))
+			//...check that they have the same balance in Abs terms and that it is not 0
+			if balanceKwithN+mapForSubK[k] != 0 && balanceKwithN != 0 {
+				log.Error(fmt.Sprintf("Expected balances to be a+b = 0 AND balance(a) != 0, but they are not, balance k with n:  %d, balance n with k: %d", balanceKwithN, mapForSubK[k]))
+				success = false
+			}
+		}
+	}
+	if !success {
+		t.Fatal("Expected balances to be symmetrical, but they were not")
+	}
+	log.Debug("test terminated")
+}
+
+//TestSwapNetworkAsymmetricFileUpload is a swap test too,
+//but this time the number and size of files are random
+func TestSwapNetworkAsymmetricFileUpload(t *testing.T) {
+	nodeCount := 16
+	balancesMap := make(map[enode.ID]map[enode.ID]int64)
+
+	sim := simulation.New(map[string]simulation.ServiceFunc{
+		"swarm": func(ctx *adapters.ServiceContext, bucket *sync.Map) (s node.Service, cleanup func(), err error) {
+			config := api.NewConfig()
+			config.Port = ""
+
+			dir, err := ioutil.TempDir("", "swap-network-test-node")
+			if err != nil {
+				return nil, nil, err
+			}
+
+			cleanup = func() {
+				err := os.RemoveAll(dir)
+				if err != nil {
+					log.Error("cleaning up swarm temp dir", "err", err)
+				}
+			}
+
+			config.Path = dir
+
+			privkey, err := crypto.GenerateKey()
+			if err != nil {
+				return nil, cleanup, err
+			}
+
+			config.Init(privkey)
+			//enable swap
+			config.SwapEnabled = true
+
+			swarm, err := NewSwarm(config, nil)
+			if err != nil {
+				return nil, cleanup, err
+			}
+			bucket.Store(bucketKeySwarm, swarm)
+			log.Info("new swarm", "bzzKey", config.BzzKey, "baseAddr", fmt.Sprintf("%x", swarm.bzz.BaseAddr()))
+			return swarm, cleanup, nil
+		},
+	})
+	defer sim.Close()
+
+	ctx := context.Background()
+	files := make([]file, 0)
+
+	//upload a snapshot
+	err := sim.UploadSnapshot(fmt.Sprintf("network/stream/testing/snapshot_%d.json", nodeCount))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	//NOTE: maxFileSize is 4 kB, this in order to provide faster tests
+	//it would be interesting to run these tests with bigger files
+	//(to see how drop limits are affected etc.)
+	const maxFileSize = 1024 * 4 //1024 bytes * 4 = 4kB
+	const minfileSize = 1024
+
+	//pseudo random algo to define if a node will upload or not
+	//if a bit is 0, do not upload
+	pseudoRandomNum := rand.Int63()
```

Comment:
let's avoid stuff like this. this is voodoo. why not just `rand.Int31() % 2 == 0`??

Reply:
that's not giving me the random I want

Reply:
i don't understand which kind of `random` do you actually need. can you please elaborate why a normal random number doesn't cut the mustard?